File: python/tests/dashboard/test_widgets.py
from .base_test import DashboardBaseTest


class TestSessions(DashboardBaseTest):

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        cls.sample_dashboard.datasources.link(cls.sample_datasource_id)

        cls.sample_widget = cls.onvoSDK.widgets.create(
            cls.sample_dashboard_id,
            "Create a pie chart showing the distribution of products by category",
        )

        cls.sample_widget_id = cls.sample_widget["id"]
        cls.sample_widget_code = cls.sample_widget["code"]

    def test_list(self):
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.widgets.list(self.sample_dashboard_id),
        )

    def test_get(self):
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.widgets.get(self.sample_widget_id),
        )

    def test_export(self):
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.widgets.export(self.sample_widget_id, "png"),
        )

    def test_update(self):
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.widgets.update(
                self.sample_widget_id, {"title": "A different title"}
            ),
        )

    def test_request_edit(self):
        sample_edit = {
            "messages": self.sample_widget["messages"]
            + [{"role": "user", "content": "Make the pie chart blue"}]
        }
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.widgets.request_edit(
                self.sample_widget_id, sample_edit
            ),
        )

    def test_execute_code(self):
        self.assertShouldRaise(
            None,
            lambda: self.onvoSDK.widgets.execute_code(
                self.sample_widget_id, self.sample_widget_code
            ),
        )

    @classmethod
    def tearDownClass(cls) -> None:
        cls.onvoSDK.widgets.delete(cls.sample_widget_id)
        super().tearDownClass()

File: ruby/test/dashboard/test_widgets.rb
# frozen_string_literal: true

require_relative './base_dashboard_template'

# All tests related to Onvo's dashboard widget endpoints (Skipping create and delete because hooks)
class WidgetTest < BaseDashboardTemplate
  def before_all
    super
    @dashboard.datasources.link(@sample_datasource_id)

    # skipping test create
    @sample_widget = @onvo.widgets.create(@sample_dashboard_id, 'Display the number of rows the data has.')
    @sample_widget_id = @sample_widget['id']
    @sample_widget_code = @sample_widget['code']
  end

  def test_list_widgets
    assert_silent { @onvo.widgets.list(@sample_dashboard_id) }
  end

  def test_get_widget
    assert_silent { @onvo.widgets.get(@sample_widget_id) }
  end

  def test_update_widget
    assert_silent { @onvo.widgets.update(@sample_widget_id, { title: 'A different title' }) }
  end

  def test_export
    assert_silent { @onvo.widgets.export(@sample_widget_id, 'png') }
  end

  def test_request_edit
    sample_edit = {
      'messages': @sample_widget['messages'] + [{ 'role': 'user', 'content': 'Make the pie chart blue' }]
    }
    assert_silent { @onvo.widgets.request_edit(@sample_widget_id, sample_edit) }
  end

  def test_execute_code
    assert_silent { @onvo.widgets.execute_code(@sample_widget_id, @sample_widget_code) }
  end

  def after_all
    super
    @onvo.widgets.delete(@sample_widget_id) # skipping test delete
  end
end
